src/repository.c

```diff
@@ -946,28 +954,48 @@ bool git_repository__reserved_names(
 }
 #endif
 
-static int check_repositoryformatversion(git_config *config)
+static int check_repositoryformatversion(int *version, git_config *config)
 {
-	int version, error;
+	int error;
 
-	error = git_config_get_int32(&version, config, "core.repositoryformatversion");
+	error = git_config_get_int32(version, config, "core.repositoryformatversion");
 	/* git ignores this if the config variable isn't there */
-	if (error == GIT_ENOTFOUND)
-		return 0;
+	if (error == GIT_ENOTFOUND) {
+		giterr_clear();
+		return GIT_REPO_VERSION;
+	}
 
 	if (error < 0)
 		return -1;
 
-	if (GIT_REPO_VERSION < version) {
+	if (*version > GIT_REPO_VERSION_MAX_ALLOWED) {
 		giterr_set(GITERR_REPOSITORY,
 			"Unsupported repository version %d. Only versions up to %d are supported.",
-			version, GIT_REPO_VERSION);
+			*version, GIT_REPO_VERSION_MAX_ALLOWED);
 		return -1;
```

Comment:
How dumb would it be to return a dedicated error code when this happens. This would be very helpful for bindings.